### PR TITLE
[SPARK-34665][SQL][DOCS] Revise the type coercion section of ANSI Compliance

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

1. Fix the table of valid type coercion combinations. Binary type should be allowed casting to String type and disallowed casting to Numeric types.
2. Summary all the `CAST`s that can cause runtime exceptions.

### Why are the changes needed?

Fix a mistake in the docs.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Run `jekyll serve` and preview:

image

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @cloud-fan @maropu 
This is to address https://github.com/apache/spark/pull/31734#discussion_r587243717
Meanwhile, I found a mistake in the table. Sorry about that.

File: docs/sql-ref-ansi-compliance.md [hunk 1]
```diff
@@ -72,16 +72,23 @@ The type conversion of Spark ANSI mode follows the syntax rules of section 6.13
 
 | Source\Target | Numeric | String | Date | Timestamp | Interval | Boolean | Binary | Array | Map | Struct |
 |-----------|---------|--------|------|-----------|----------|---------|--------|-------|-----|--------|
-| Numeric   | Y       | Y      | N    | N         | N        | Y       | N      | N     | N   | N      |
-| String    | Y       | Y      | Y    | Y         | Y        | Y       | Y      | N     | N   | N      |
+| Numeric   | <span style="color:red">Y</span> | Y      | N    | N         | N        | Y       | N      | N     | N   | N      |
```
Comment (maropu) [inline docs/sql-ref-ansi-compliance.md:75]: For following the universal design, could we use a bold char or `*` instead? Or use the combination of them?

File: docs/sql-ref-ansi-compliance.md [hunk 2]
```diff
@@ -72,16 +72,23 @@ The type conversion of Spark ANSI mode follows the syntax rules of section 6.13
...
-| Array     | N       | N      | N    | N         | N        | N       | N      | Y     | N   | N      |
-| Map       | N       | N      | N    | N         | N        | N       | N      | N     | Y   | N      |
-| Struct    | N       | N      | N    | N         | N        | N       | N      | N     | N   | Y      |
+| Binary    | N       | Y      | N    | N         | N        | N       | Y      | N     | N   | N      |
+| Array     | N       | N      | N    | N         | N        | N       | N      | <span style="color:red">Y</span> | N   | N      |
+| Map       | N       | N      | N    | N         | N        | N       | N      | N     | <span style="color:red">Y</span> | N      |
+| Struct    | N       | N      | N    | N         | N        | N       | N      | N     | N   | <span style="color:red">Y</span> |
+
+In the table above, all the `CAST`s that can cause runtime exceptions are marked as red <span style="color:red">Y</span>:
+* CAST(Numeric AS Numeric): raise a overflow exception if the value is out of the target data type's range.
```
Comment (maropu) [inline docs/sql-ref-ansi-compliance.md:87]: `a overflow` -> `an overflow`

**Review by maropu — APPROVED** [review]
I left minor comments and LGTM otherwise.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by maropu** [comment]
Thanks for the update, @gengliangwang ! LGTM and I'll merge this.

**Comment by cloud-fan** [comment]
late LGTM

**Pull request closed without merging**